### PR TITLE
Replace `&vec[0]` with `vec.data()`

### LINKED ISSUE #6599: Replace `&vec[0]` with `vec.data()`
This is just more idiomatic, and more important, won't fail on empty vectors.
Right now we unconditionally call this on possibly-empty vectors.
For example:
https://github.com/zeroc-ice/ice/blob/354a325fea7a0cb0dc316d8730c9f93cdd16ce23/cpp/src/Ice/SSL/SecureTransportUtil.cpp#L613

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This fixes #6599.
This is primarily about code-hardening and better intent-signaling, but this also fixes a couple of bugs (albeit uncommon ones)

- UB in Python, Ruby, and MATLAB: if you have an opaque endpoint with a header-only encapsulation and call `getInfo()`.
- UB on MacOS if you load a bad PEM certificate with only whitespace between `BEGIN` and `END` markers.
- UB in other obscure cert related code in the bowels of IceSSL. These also require bad certificates though.

There were also places where we were doing `&xxx[0]` redundantly. These were just outright removed.

### CODE REVIEW

File: python/modules/IcePy/EndpointInfo.cpp [hunk 1]
```diff
@@ -155,7 +155,7 @@ opaqueEndpointInfoGetRawBytes(EndpointInfoObject* self, PyObject* /*args*/)
     auto info = dynamic_pointer_cast<Ice::OpaqueEndpointInfo>(*self->endpointInfo);
     assert(info);
     return PyBytes_FromStringAndSize(
-        reinterpret_cast<const char*>(&info->rawBytes[0]),
+        reinterpret_cast<const char*>(info->rawBytes.data()),
         static_cast<int>(info->rawBytes.size()));
```
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply python/modules/IcePy/EndpointInfo.cpp:159]: I don't think people are ever encountering header-only opaque endpoints.

File: cpp/src/Ice/SSL/SecureTransportUtil.cpp [hunk 2]
```diff
@@ -610,7 +610,7 @@ namespace
                 vector<byte> data(IceInternal::Base64::decode(string(&buffer[startpos], size)));
                 UniqueRef<CFDataRef> certdata(CFDataCreate(
                     kCFAllocatorDefault,
-                    reinterpret_cast<uint8_t*>(&data[0]),
+                    reinterpret_cast<uint8_t*>(data.data()),
                     static_cast<CFIndex>(data.size())));
```
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply cpp/src/Ice/SSL/SecureTransportUtil.cpp:625]: I think this is obscure enough to not need a test.
Whitespace only certificates are just too generate to care about.

File: matlab/src/Endpoint.cpp [hunk 3]
```diff
@@ -101,7 +101,7 @@ namespace
                 {
                     // We don't set InfoType because there is no reserved int16 constant for Opaque.
                     mxSetFieldByNumber(r, 0, Field::RawEncoding, createEncodingVersion(opaqueInfo->rawEncoding));
-                    const std::byte* p = &opaqueInfo->rawBytes[0];
+                    const std::byte* p = opaqueInfo->rawBytes.data();
                     mxSetFieldByNumber(r, 0, Field::RawBytes, createByteArray(p, p + opaqueInfo->rawBytes.size()));
```
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply matlab/src/Endpoint.cpp:105]: This is slightly inaccurate, C++ fully allows you to do `nullptr + 0` (just equals `nullptr`), so there's nothing wrong with that part of the logic. What was wrong is that we were passing `nullptr` into `memcpy`. In practice, harmless since this only happened when the length was `0` and a `memcpy` of `length = 0` is no-op, regardless of pointers.

But, I still fixed this where the actual bug lies (`matlab/src/Util.cpp`), and we don't call the `memcpy` when we have `nullptr`.

File: python/modules/IcePy/EndpointInfo.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply python/modules/IcePy/EndpointInfo.cpp:159]: There is simply no easy way to test this. The only way to construct a header-only opaque endpoint is by hand-crafting the bytes we decode from. This is A) really hard to do with our test harness, and B) really brittle, and C) looks utterly magical from the outside.

This is enough of an edge case, that we'll just omit testing it.

File: ruby/src/IceRuby/Endpoint.cpp [hunk 4]
```diff
@@ -176,7 +176,7 @@ IceRuby::createEndpointInfo(const Ice::EndpointInfoPtr& p)
 
         Ice::OpaqueEndpointInfoPtr opaque = dynamic_pointer_cast<Ice::OpaqueEndpointInfo>(p);
         auto b = opaque->rawBytes;
-        volatile VALUE v = callRuby(rb_str_new, reinterpret_cast<const char*>(&b[0]), static_cast<long>(b.size()));
+        volatile VALUE v = callRuby(rb_str_new, reinterpret_cast<const char*>(b.data()), static_cast<long>(b.size()));
```
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply ruby/src/IceRuby/Endpoint.cpp:179]: See https://github.com/zeroc-ice/ice/pull/6604/#discussion_r3787416319
